### PR TITLE
DRAFT: Add logging to all method calls

### PR DESCRIPTION
**Pull request by jbrundage**
@GordonSmith Take a look at this concept. This would allow us to track the full life cycle of a widget by hooking into every method call. Writing to the console is just one example of how this could be used... but imagine the start and end log statements as the timestamp ranges. The method calls could be drawn to the MiniGantt... and their ranges could be colored based on which outer method they were called from (enter/update/exit).

### CODE REVIEW

File: packages/common/src/Widget.ts [hunk 1]
```diff
@@ -66,10 +67,13 @@ export abstract class Widget extends PropertyExt {
 
     protected _overlayElement;
 
+    protected _logger;
+
     constructor() {
         super();
         this._class = Object.getPrototypeOf(this)._class;
         this._id = this._idSeed + widgetID++;
+        this._logger = scopedLogger(this._id);
```
Comment (GordonSmith) [inline packages/common/src/Widget.ts]: Interesting - I only envisioned using the scoping for "per file" not "per widget instance"

File: packages/common/src/Widget.ts [hunk 2]
```diff
@@ -657,7 +661,21 @@ export abstract class Widget extends PropertyExt {
         return this;
     }
 
-    enter(_domNode: HTMLElement, _element: d3SelectionType) { }
+    enter(_domNode: HTMLElement, _element: d3SelectionType) {
+        if ((window as any).__hpcc_debug) {
+            const context = this;
```
Comment (GordonSmith) [inline packages/common/src/Widget.ts]: No need for context

File: packages/common/src/Widget.ts [hunk 3]
```diff
@@ -657,7 +661,21 @@ export abstract class Widget extends PropertyExt {
         return this;
     }
 
-    enter(_domNode: HTMLElement, _element: d3SelectionType) { }
+    enter(_domNode: HTMLElement, _element: d3SelectionType) {
+        if ((window as any).__hpcc_debug) {
```
Comment (GordonSmith) [inline packages/common/src/Widget.ts]: I think this code would need its own "switch" as its going to be very very noisy. 
Also it should be done on the constructor and probably in the `Class` class?

Comment (GordonSmith) [reply packages/common/src/Widget.ts]: And maybe only for specific methods?

Comment (jbrundage) [reply packages/common/src/Widget.ts]: Hmm... maybe another global variable could be used to control which methods are logged?
For example:
```javascript
window.__hpcc_debug_methods = ["enter","update","exit","render","resize","target","columns","data"];
```
...and how they're logged:
```javascript
window.__hpcc_debug_callbacks = {
    "args": methodName=>{
        this._logger.args(methodName, arguments);
    },
    "timestamps": methodName=>{
        this._logger.timestamps(methodName, Date.getTime());
    },
    "trace": methodName=>{
        console.groupCollapsed(this._logger._id+" "+methodName);
        console.trace();
        console.groupEnd();
    }
};
```

**Comment by jbrundage** [comment]
@GordonSmith what's the take-away from this? It would be very handy to be able to modify a global debug config to tell it to log specific methods (particularly when debugging issues in production)

File: packages/common/src/Widget.ts [hunk 4]
```diff
@@ -657,7 +661,21 @@ export abstract class Widget extends PropertyExt {
...
+                const orig = context[methodName];
+                context[methodName] = function() {
+                    context._logger.debug(`start ${methodName}`);
+                    const ret = orig.apply(this, arguments);
+                    context._logger.debug(`end ${methodName}`);
+                    return ret;
+                };
+            });
+        }
+    }
```
Comment (GordonSmith) [inline packages/common/src/Widget.ts]: This code should really be in Class.ts

**Pull request closed without merging**